File: numberOfPaths/numberOfPaths/src/main/java/com/mycompany/numberofpaths/Tabulation.java
package com.mycompany.numberofpaths;

public class Tabulation {
    
    /**
     * @param matrix - matriz a la que se le aplicará el problema de tabulation
     * @param givenCost - coste a buscar
    */
    public static void findNumberOfPathsTabulation(int[][] matrix, int givenCost) {
        long start = System.nanoTime();

        int rows = matrix.length;
        int cols = matrix[0].length;
        
        int [][][]tab = new int[rows][cols][givenCost + 1];
        
        if (matrix[0][0] > givenCost) Data.setPaths(0);
        tab[0][0][matrix[0][0]] = 1;

        int currentCost = 0;
        for (int i = 0; i < rows; i++) {
            currentCost += matrix[i][0];
            if (currentCost <= givenCost) tab[i][0][currentCost] = 1;
        }

        currentCost = 0;
        for (int i = 0; i < cols; i++) {
            currentCost += matrix[0][i];
            if (currentCost <= givenCost) tab[0][i][currentCost] = 1;
        }

        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                for (int k = 0; k <= givenCost; k++) {
                    if (k - matrix[i][j] >= 0) tab[i][j][k] = tab[i - 1][j][k - matrix[i][j]] + tab[i][j - 1][k - matrix[i][j]];
                }
            }
        }

        Data.setPaths(tab[rows - 1][cols - 1][givenCost]);
        Data.setTime(String.format("%.2f", (double) (System.nanoTime() - start)/1_000_000_000));
    }
}

File: numberOfPath-Python/tabulation.py
def find_number_of_paths_tabulation(matrix, info):
    given_cost = info.get_cost()
    rows = len(matrix)
    cols = len(matrix[0])

    tab = [[[0 for i in range(int(given_cost) + 1)] for j in range(cols)] for k in range(rows)]

    if int(matrix[0][0]) > int(given_cost):
        info.set_path(0)

    tab[0][0][int(matrix[0][0])] = 1

    current_cost = 0
    for i in range(rows):
        current_cost += int(matrix[i][0])
        if current_cost <= int(given_cost):
            tab[i][0][current_cost] = 1

    current_cost = 0
    for i in range(cols):
        current_cost += int(matrix[0][i])
        if current_cost <= int(given_cost):
            tab[0][i][current_cost] = 1

    for i in range(1, rows):
        for j in range(1, cols):
            for k in range(int(given_cost) + 1):
                if k - int(matrix[i][j]) >= 0:
                    tab[i][j][k] = tab[i - 1][j][k - int(matrix[i][j])] + tab[i][j - 1][k - int(matrix[i][j])]

    info.set_path(tab[rows - 1][cols - 1][int(given_cost)])

    return info
